File: native-schema-registry/golang/integration-tests/pkg/fakeglue/fake_glue.go
//go:build integration

// Package fakeglue provides a minimal in-memory fake of the
// gsrcore.GlueClient interface. The integration-tests module uses it
// to drive the Phase 4 lifecycle / cache / error scenarios (§5.3
// items 13-29) without billing real Glue.
//
// fakeglue is NOT a replacement for the testify/mock fake under
// pkg/gsrserde-go/core/ (which is _test.go-only). Reach for testify/
// mock when a test needs call-order or argument-matcher assertions;
// reach for fakeglue when the test wants stateful schema-by-
// definition / schema-by-version-id lookups that mimic the real Glue
// API at a behavioral level.
package fakeglue

import (
	"context"
	"fmt"
	"sync"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/glue"
	"github.com/aws/aws-sdk-go-v2/service/glue/types"
	"github.com/google/uuid"
)

// Fake is an in-memory GlueClient. It implements the schema-by-
// definition / schema-by-version-id lookup behaviors that the core
// encoder/decoder rely on, plus enough of the metadata + tag APIs to
// satisfy the GlueClient interface.
//
// Concurrency: methods are goroutine-safe; tests can drive Encode
// from N goroutines without external locking.
type Fake struct {
	mu sync.Mutex

	// definitions maps (registryName + "/" + schemaName + "/" +
	// schemaDefinition) -> schemaVersionId, mimicking the
	// GetSchemaByDefinition lookup key.
	definitions map[string]string

	// versions maps schemaVersionId -> stored schema.
	versions map[string]*storedSchema

	// CallCounts is a counter map keyed by Glue API method name.
	// Prefer Count(name) / Snapshot() over reading this map directly:
	// writes go under f.mu, so a direct mid-flight read would race
	// (the field is exported only so legacy tests that read it after
	// wg.Wait() keep compiling — those reads are safe via the
	// happens-before edge wg.Wait gives, but the raw map invites
	// future misuse).
	CallCounts map[string]int

	// ForceCreateError, when non-nil, makes CreateSchema return this
	// error verbatim. Used by negative scenarios (§5.3 items 23-25).
	ForceCreateError error

	// ForceGetSchemaError, when non-nil, makes GetSchemaByDefinition
	// return this error verbatim. Used to simulate IAM denied / Throttling.
	ForceGetSchemaError error

	// ForceGetVersionError, when non-nil, makes GetSchemaVersion
	// return this error verbatim.
	ForceGetVersionError error

	// ForceRegisterPending, when true, makes RegisterSchemaVersion return
	// Status=PENDING instead of AVAILABLE. Pair with ForcePendingCount to
	// script a "N PENDING responses then AVAILABLE" sequence via the
	// GetSchemaVersion poll loop — mirroring the Tier-1 mock approach in
	// poll_test.go for integration-level Tier-2 scenarios (spec §5.2 PBI-11).
	ForceRegisterPending bool

	// ForcePendingCount is the number of GetSchemaVersion calls that will
	// return Status=PENDING before switching to AVAILABLE. Decremented on
	// each call. When it reaches zero, subsequent GetSchemaVersion calls
	// return the normal AVAILABLE response (unless ForceGetVersionError is
	// set). Only meaningful when ForceRegisterPending=true triggers the poll.
	ForcePendingCount int
}

type storedSchema struct {
	registryName     string
	schemaName       string
	schemaDefinition string
	dataFormat       types.DataFormat
}

// New returns an empty Fake.
func New() *Fake {
	return &Fake{
		definitions: make(map[string]string),
		versions:    make(map[string]*storedSchema),
		CallCounts:  make(map[string]int),
	}
}

func (f *Fake) bump(name string) {
	f.CallCounts[name]++
}

// Count returns the call count for the given Glue API method under
// the mutex. Use this from tests that may read mid-flight; direct
// access to f.CallCounts is safe only after a happens-before edge
// (e.g. wg.Wait).
func (f *Fake) Count(name string) int {
	f.mu.Lock()
	defer f.mu.Unlock()
	return f.CallCounts[name]
}

// Snapshot returns a copy of CallCounts taken under the mutex.
func (f *Fake) Snapshot() map[string]int {
	f.mu.Lock()
	defer f.mu.Unlock()
	out := make(map[string]int, len(f.CallCounts))
	for k, v := range f.CallCounts {
		out[k] = v
	}
	return out
}

// definitionKey returns the in-memory lookup key for
// (registry, schema-name, schema-definition). registryName may be ""
// (the Java default) which deliberately maps to a distinct key from
// "default-registry" so callers see whatever the encoder actually
// sent over.
func definitionKey(registry, name, def string) string {
	return fmt.Sprintf("%s/%s/%s", registry, name, def)
}

// GetSchemaByDefinition implements gsrcore.GlueClient.
func (f *Fake) GetSchemaByDefinition(ctx context.Context, in *glue.GetSchemaByDefinitionInput, _ ...func(*glue.Options)) (*glue.GetSchemaByDefinitionOutput, error) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.bump("GetSchemaByDefinition")

	if f.ForceGetSchemaError != nil {
		return nil, f.ForceGetSchemaError
	}

	registry := ""
	name := ""
	if in.SchemaId != nil {
		if in.SchemaId.RegistryName != nil {
			registry = *in.SchemaId.RegistryName
		}
		if in.SchemaId.SchemaName != nil {
			name = *in.SchemaId.SchemaName
		}
	}
	def := ""
	if in.SchemaDefinition != nil {
		def = *in.SchemaDefinition
	}

	versionID, ok := f.definitions[definitionKey(registry, name, def)]
	if !ok {
		return nil, &types.EntityNotFoundException{Message: aws.String("schema not found")}
	}
	return &glue.GetSchemaByDefinitionOutput{
		SchemaVersionId: aws.String(versionID),
		Status:          types.SchemaVersionStatusAvailable,
	}, nil
}

// GetSchemaVersion implements gsrcore.GlueClient.
func (f *Fake) GetSchemaVersion(ctx context.Context, in *glue.GetSchemaVersionInput, _ ...func(*glue.Options)) (*glue.GetSchemaVersionOutput, error) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.bump("GetSchemaVersion")

	if f.ForceGetVersionError != nil {
		return nil, f.ForceGetVersionError
	}

	// Script "N PENDING then AVAILABLE" for poll-path Tier-2 tests (spec §5.2
	// PBI-11). The counter is decremented under the mutex so concurrent callers
	// each see the correct sequence without races.
	if f.ForcePendingCount > 0 {
		f.ForcePendingCount--
		id := ""
		if in.SchemaVersionId != nil {
			id = *in.SchemaVersionId
		}
		return &glue.GetSchemaVersionOutput{
			SchemaVersionId: aws.String(id),
			Status:          types.SchemaVersionStatusPending,
		}, nil
	}

	id := ""
	if in.SchemaVersionId != nil {
		id = *in.SchemaVersionId
	}
	stored, ok := f.versions[id]
	if !ok {
		return nil, &types.EntityNotFoundException{Message: aws.String("schema version not found")}
	}
	arn := fmt.Sprintf("arn:aws:glue:us-east-1:0:schema/%s/%s", stored.registryName, stored.schemaName)
	return &glue.GetSchemaVersionOutput{
		SchemaVersionId:  aws.String(id),
		SchemaDefinition: aws.String(stored.schemaDefinition),
		DataFormat:       stored.dataFormat,
		SchemaArn:        aws.String(arn),
		Status:           types.SchemaVersionStatusAvailable,
	}, nil
}

// CreateSchema implements gsrcore.GlueClient.
func (f *Fake) CreateSchema(ctx context.Context, in *glue.CreateSchemaInput, _ ...func(*glue.Options)) (*glue.CreateSchemaOutput, error) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.bump("CreateSchema")

	if f.ForceCreateError != nil {
		return nil, f.ForceCreateError
	}

	registry := ""
	if in.RegistryId != nil && in.RegistryId.RegistryName != nil {
		registry = *in.RegistryId.RegistryName
	}
	name := ""
	if in.SchemaName != nil {
		name = *in.SchemaName
	}
	def := ""
	if in.SchemaDefinition != nil {
		def = *in.SchemaDefinition
	}

	versionID := uuid.NewString()
	f.definitions[definitionKey(registry, name, def)] = versionID
	f.versions[versionID] = &storedSchema{
		registryName:     registry,
		schemaName:       name,
		schemaDefinition: def,
		dataFormat:       in.DataFormat,
	}

	v := int64(1)
	return &glue.CreateSchemaOutput{
		SchemaVersionId:     aws.String(versionID),
		LatestSchemaVersion: &v,
	}, nil
}

// RegisterSchemaVersion implements gsrcore.GlueClient.
//
// Real Glue derives DataFormat from the parent schema (registered via
// CreateSchema); Register doesn't carry it on the input. To match that
// behavior the fake inherits DataFormat from any prior storedSchema
// under (registry, name). Without this inheritance, the decoder side
// would receive DataFormat="" on the AlreadyExists fall-through path
// and silently skip protobuf message-index stripping (decoder.go:75),
// corrupting the decoded bytes.
func (f *Fake) RegisterSchemaVersion(ctx context.Context, in *glue.RegisterSchemaVersionInput, _ ...func(*glue.Options)) (*glue.RegisterSchemaVersionOutput, error) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.bump("RegisterSchemaVersion")

	versionID := uuid.NewString()
	registry := ""
	name := ""
	if in.SchemaId != nil {
		if in.SchemaId.RegistryName != nil {
			registry = *in.SchemaId.RegistryName
		}
		if in.SchemaId.SchemaName != nil {
			name = *in.SchemaId.SchemaName
		}
	}
	def := ""
	if in.SchemaDefinition != nil {
		def = *in.SchemaDefinition
	}
	f.definitions[definitionKey(registry, name, def)] = versionID

	// Inherit DataFormat from any prior storedSchema for this
	// (registry, name) — mirrors real Glue. If no prior schema exists
	// the format stays zero-value, matching real Glue's behavior of
	// rejecting RegisterSchemaVersion against an unknown schema name.
	var inheritedFormat types.DataFormat
	for _, prior := range f.versions {
		if prior.registryName == registry && prior.schemaName == name && prior.dataFormat != "" {
			inheritedFormat = prior.dataFormat
			break
		}
	}
	f.versions[versionID] = &storedSchema{
		registryName:     registry,
		schemaName:       name,
		schemaDefinition: def,
		dataFormat:       inheritedFormat,
	}

	v := int64(2)
	status := types.SchemaVersionStatusAvailable
	if f.ForceRegisterPending {
		status = types.SchemaVersionStatusPending
	}
	return &glue.RegisterSchemaVersionOutput{
		SchemaVersionId: aws.String(versionID),
		VersionNumber:   &v,
		Status:          status,
	}, nil
}

// PutSchemaVersionMetadata implements gsrcore.GlueClient.
func (f *Fake) PutSchemaVersionMetadata(ctx context.Context, in *glue.PutSchemaVersionMetadataInput, _ ...func(*glue.Options)) (*glue.PutSchemaVersionMetadataOutput, error) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.bump("PutSchemaVersionMetadata")
	return &glue.PutSchemaVersionMetadataOutput{}, nil
}

// QuerySchemaVersionMetadata implements gsrcore.GlueClient.
func (f *Fake) QuerySchemaVersionMetadata(ctx context.Context, in *glue.QuerySchemaVersionMetadataInput, _ ...func(*glue.Options)) (*glue.QuerySchemaVersionMetadataOutput, error) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.bump("QuerySchemaVersionMetadata")
	return &glue.QuerySchemaVersionMetadataOutput{}, nil
}

// GetTags implements gsrcore.GlueClient.
func (f *Fake) GetTags(ctx context.Context, in *glue.GetTagsInput, _ ...func(*glue.Options)) (*glue.GetTagsOutput, error) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.bump("GetTags")
	return &glue.GetTagsOutput{}, nil
}

File: multilang-schema-registry/typescript/packages/integration-tests/src/fake-glue.ts
/**
 * In-memory `GlueClient` fake for Tier-2 tests.
 *
 * `FakeGlueClient` implements the `@gsr/core` `GlueClient` seam entirely in
 * process — no network, no Docker, no AWS. It maintains a definition→versionId
 * lookup and a versionId→stored-schema map, mirrors Go's `fakeglue.Fake` beat
 * for beat, and exposes:
 *
 *   - Per-method `callCounts` for de-dup / cache-hit assertions.
 *   - `Force*` affordances that let a scenario simulate any of the four Glue
 *     failure modes the classifier discriminates (`EntityNotFoundException`,
 *     `AlreadyExistsException`, `ThrottlingException`, `AccessDeniedException`)
 *     as well as the async register→poll convergence.
 *   - Named-error factory helpers producing `.name`-discriminated errors that
 *     `@gsr/core.classifyGlueError` reads exactly the way it reads
 *     AWS SDK v3 service errors (via the outer error's `.name`).
 *
 * The fake is a Tier-2 substitute; where a scenario needs call-order or
 * argument-matcher assertions it may layer `aws-sdk-client-mock` against the
 * concrete SDK client instead. Both mechanisms are permitted; a scenario
 * picks whichever expresses its assertion most directly.
 */
import type {
  CreateSchemaCommandInput,
  CreateSchemaCommandOutput,
  GetSchemaByDefinitionCommandInput,
  GetSchemaByDefinitionCommandOutput,
  GetSchemaVersionCommandInput,
  GetSchemaVersionCommandOutput,
  GetTagsCommandInput,
  GetTagsCommandOutput,
  PutSchemaVersionMetadataCommandInput,
  PutSchemaVersionMetadataCommandOutput,
  QuerySchemaVersionMetadataCommandInput,
  QuerySchemaVersionMetadataCommandOutput,
  RegisterSchemaVersionCommandInput,
  RegisterSchemaVersionCommandOutput,
} from "@aws-sdk/client-glue";
import type { GlueClient } from "@gsr/core";

/**
 * The seven method names the fake tracks call counts for. Matches
 * `@gsr/core` `GlueClient` one-to-one.
 */
export type FakeGlueMethodName =
  | "getSchemaByDefinition"
  | "getSchemaVersion"
  | "createSchema"
  | "registerSchemaVersion"
  | "putSchemaVersionMetadata"
  | "querySchemaVersionMetadata"
  | "getTags";

/**
 * Zero-initialised call-count map. Kept as a top-level constant so a fresh
 * counter can be produced on construction and on `reset()` without listing
 * every method name in-line each time.
 */
const ZERO_CALL_COUNTS: Readonly<Record<FakeGlueMethodName, number>> =
  Object.freeze({
    getSchemaByDefinition: 0,
    getSchemaVersion: 0,
    createSchema: 0,
    registerSchemaVersion: 0,
    putSchemaVersionMetadata: 0,
    querySchemaVersionMetadata: 0,
    getTags: 0,
  });

/**
 * Named-error `.name` values the fake can raise on the `Force*` fields and
 * that consumers may throw from their own paths. Wired to
 * `@gsr/core.classifyGlueError`, which discriminates on `err.name` exactly
 * the way AWS SDK v3 service errors expose it.
 */
export const GLUE_ERROR_NAMES = {
  EntityNotFound: "EntityNotFoundException",
  AlreadyExists: "AlreadyExistsException",
  Throttling: "ThrottlingException",
  AccessDenied: "AccessDeniedException",
} as const;

/**
 * Build an `Error` whose `.name` matches an AWS SDK v3 Glue service error so
 * `classifyGlueError` (and any `.name`-based comparison) treats it as the
 * real thing. AWS SDK v3 always sets `.name` on service errors to the exact
 * Smithy shape name; setting `.name` on a plain `Error` reproduces that
 * discriminant faithfully without pulling the SDK into this file's runtime
 * dependency graph.
 */
function makeNamedError(name: string, message: string): Error {
  const err = new Error(message);
  err.name = name;
  return err;
}

/**
 * `EntityNotFoundException` factory — the auto-register fall-through signal.
 * Raised by `GetSchemaByDefinition` when the definition is not yet known to
 * Glue; the registrar branches on this to run `CreateSchema`.
 */
export function makeEntityNotFoundError(message = "entity not found"): Error {
  return makeNamedError(GLUE_ERROR_NAMES.EntityNotFound, message);
}

/**
 * `AlreadyExistsException` factory — the concurrent-producer race signal.
 * Raised by `CreateSchema` when a sibling producer registered the same
 * schema first; the registrar falls through to `RegisterSchemaVersion`.
 */
export function makeAlreadyExistsError(message = "already exists"): Error {
  return makeNamedError(GLUE_ERROR_NAMES.AlreadyExists, message);
}

/**
 * `ThrottlingException` factory. Not on the register-flow branching path
 * (surfaces as `other` in `classifyGlueError`) but exposed here so scenarios
 * can drive backoff/retry surfacing.
 */
export function makeThrottlingError(message = "rate exceeded"): Error {
  return makeNamedError(GLUE_ERROR_NAMES.Throttling, message);
}

/**
 * `AccessDeniedException` factory. Surfaces IAM-denied paths as a client
 * error; classifies as `other` and therefore propagates as
 * `GsrRegistrationError` when caught on the register flow.
 */
export function makeAccessDeniedError(message = "access denied"): Error {
  return makeNamedError(GLUE_ERROR_NAMES.AccessDenied, message);
}

/**
 * Options accepted at construction time. Every field is a Tier-2
 * determinism seam — none is required for a bare `new FakeGlueClient()`.
 */
export interface FakeGlueOptions {
  /**
   * `SchemaVersionId` UUIDs the fake will hand out on `CreateSchema` /
   * `RegisterSchemaVersion`, in order. Callers who assert on specific UUIDs
   * supply their own list; when omitted, the fake generates deterministic
   * UUIDs (`00000000-0000-0000-0000-<12-hex>` where the low bytes are a
   * monotonic counter) so the test output is reproducible without depending
   * on `crypto.randomUUID`.
   */
  versionIds?: string[];
}

/**
 * Internal shape stored per registered schema version. `SchemaDefinition` +
 * `DataFormat` are what a caller sends on `RegisterSchemaVersion` /
 * `CreateSchema`; the fake retains them so `GetSchemaVersion` can return
 * them and so a duplicate-definition lookup on `GetSchemaByDefinition`
 * resolves without re-registering.
 */
interface StoredSchemaVersion {
  schemaVersionId: string;
  schemaName: string;
  registryName: string;
  schemaDefinition: string;
  dataFormat: string;
  compatibility?: string;
  description?: string;
  tags: Record<string, string>;
  metadata: Record<string, string>;
  /** Poll-count remaining before this version flips `AVAILABLE`. */
  pendingCount: number;
}

/**
 * Compose the definition-lookup key used by `GetSchemaByDefinition` and
 * (internally) by `CreateSchema`. Registry + schema name + full definition
 * so two schemas with different names but identical bodies never collide,
 * and so a caller who repeats the same (name, definition) lookup gets the
 * previously registered version back.
 */
function definitionKey(
  registryName: string | undefined,
  schemaName: string | undefined,
  schemaDefinition: string | undefined,
): string {
  return `${registryName ?? ""}|${schemaName ?? ""}|${schemaDefinition ?? ""}`;
}

/**
 * In-memory Glue fake. Implements the full `GlueClient` seam interface so
 * TypeScript compile-checks any drift; a scenario constructs a bare
 * `new FakeGlueClient()`, tweaks the `Force*` fields per test, and hands
 * the instance to the registrar/metadata reader.
 *
 * Not thread-safe (Node is single-threaded per event loop; concurrent
 * `Promise` fan-out on one instance is fine because the fake's writes are
 * atomic within a microtask).
 */
export class FakeGlueClient implements GlueClient {
  /**
   * Definition→versionId lookup keyed by
   * `${registryName}|${schemaName}|${schemaDefinition}`.
   */
  private readonly definitionToVersion = new Map<string, string>();

  /** versionId→stored-schema map. */
  private readonly versionToSchema = new Map<string, StoredSchemaVersion>();

  /** Per-method call counter (public via {@link callCounts}). */
  private readonly counts: Record<FakeGlueMethodName, number> = {
    ...ZERO_CALL_COUNTS,
  };

  /** Preseeded UUIDs from the constructor options, popped left-to-right. */
  private readonly seededVersionIds: string[];

  /** Monotonic counter for the fallback UUID generator. */
  private nextGeneratedId = 1;

  // --- Force* affordances -------------------------------------------------

  /**
   * When set, `getSchemaByDefinition` throws this error instead of consulting
   * the store. Set to a `makeEntityNotFoundError()` (or any classifier-typed
   * error) to drive control-flow branches.
   */
  public forceGetSchemaError: Error | null = null;

  /**
   * When set, `createSchema` throws this error. Set to
   * `makeAlreadyExistsError()` to drive the concurrent-producer race path.
   */
  public forceCreateError: Error | null = null;

  /**
   * When set, `getSchemaVersion` throws this error. Drives poll-path
   * failure.
   */
  public forceGetVersionError: Error | null = null;

  /**
   * When true, a newly created/registered schema version starts in
   * `PENDING` state and requires {@link forcePendingCount} `GetSchemaVersion`
   * calls before flipping to `AVAILABLE`. When false, versions are
   * `AVAILABLE` immediately.
   */
  public forceRegisterPending = false;

  /**
   * Number of `PENDING` responses `GetSchemaVersion` returns before the
   * version flips to `AVAILABLE`. Only consulted when
   * {@link forceRegisterPending} is true. `0` means "flip on the first
   * poll" (already `AVAILABLE`); larger values force N `PENDING` polls
   * then one `AVAILABLE`.
   */
  public forcePendingCount = 0;

  constructor(opts: FakeGlueOptions = {}) {
    this.seededVersionIds = opts.versionIds ? [...opts.versionIds] : [];
  }

  /**
   * Per-method call counter. Read-only view — mutations happen inside the
   * seam methods only. Returned as a shallow copy so a caller who snapshots
   * this value at two points can diff them without racing the internal
   * counter.
   */
  get callCounts(): Readonly<Record<FakeGlueMethodName, number>> {
    return { ...this.counts };
  }

  /**
   * Reset counters, `Force*` fields, and the schema store to the pristine
   * post-construction state. Useful in `beforeEach`.
   */
  reset(): void {
    this.definitionToVersion.clear();
    this.versionToSchema.clear();
    for (const k of Object.keys(this.counts) as FakeGlueMethodName[]) {
      this.counts[k] = 0;
    }
    this.forceGetSchemaError = null;
    this.forceCreateError = null;
    this.forceGetVersionError = null;
    this.forceRegisterPending = false;
    this.forcePendingCount = 0;
    this.nextGeneratedId = 1;
  }

  /**
   * Directly read a stored version — for test-side assertions that a
   * scenario wrote what it thought it wrote. Not part of the `GlueClient`
   * seam.
   */
  peekVersion(schemaVersionId: string): StoredSchemaVersion | undefined {
    return this.versionToSchema.get(schemaVersionId);
  }

  // --- GlueClient seam methods -------------------------------------------

  async getSchemaByDefinition(
    input: GetSchemaByDefinitionCommandInput,
  ): Promise<GetSchemaByDefinitionCommandOutput> {
    this.counts.getSchemaByDefinition += 1;
    if (this.forceGetSchemaError) {
      throw this.forceGetSchemaError;
    }

    const key = definitionKey(
      input.SchemaId?.RegistryName,
      input.SchemaId?.SchemaName,
      input.SchemaDefinition,
    );
    const versionId = this.definitionToVersion.get(key);
    if (versionId === undefined) {
      // Not-found is expressed as a typed error (matching real Glue and
      // `classifyGlueError`), not as a null response — the registrar
      // branches on `EntityNotFoundException`, so returning `undefined`
      // would silently break the auto-register flow.
      throw makeEntityNotFoundError(
        `no schema version found for schemaName=${
          input.SchemaId?.SchemaName ?? ""
        }`,
      );
    }
    const stored = this.versionToSchema.get(versionId);
    return {
      $metadata: {},
      SchemaVersionId: versionId,
      Status: this.statusFor(stored),
      SchemaArn: this.schemaArnFor(stored),
      DataFormat: stored?.dataFormat as
        | GetSchemaByDefinitionCommandOutput["DataFormat"]
        | undefined,
    };
  }

  async getSchemaVersion(
    input: GetSchemaVersionCommandInput,
  ): Promise<GetSchemaVersionCommandOutput> {
    this.counts.getSchemaVersion += 1;
    if (this.forceGetVersionError) {
      throw this.forceGetVersionError;
    }

    const versionId = input.SchemaVersionId;
    if (versionId === undefined) {
      throw makeEntityNotFoundError("SchemaVersionId is required");
    }
    const stored = this.versionToSchema.get(versionId);
    if (stored === undefined) {
      throw makeEntityNotFoundError(
        `no schema version found: schemaVersionId=${versionId}`,
      );
    }

    // Poll-loop convergence: each `getSchemaVersion` call decrements
    // `pendingCount` until it hits zero, at which point the caller sees
    // `AVAILABLE`. Modelled as a countdown so `forcePendingCount=2` matches
    // "return PENDING twice, then AVAILABLE" — Go parity with the async
    // poll-to-AVAILABLE loop.
    let status: "AVAILABLE" | "PENDING";
    if (stored.pendingCount > 0) {
      status = "PENDING";
      stored.pendingCount -= 1;
    } else {
      status = "AVAILABLE";
    }

    return {
      $metadata: {},
      SchemaVersionId: versionId,
      Status: status,
      SchemaArn: this.schemaArnFor(stored),
      SchemaDefinition: stored.schemaDefinition,
      DataFormat: stored.dataFormat as
        | GetSchemaVersionCommandOutput["DataFormat"]
        | undefined,
    };
  }

  async createSchema(
    input: CreateSchemaCommandInput,
  ): Promise<CreateSchemaCommandOutput> {
    this.counts.createSchema += 1;
    if (this.forceCreateError) {
      throw this.forceCreateError;
    }

    const registryName = input.RegistryId?.RegistryName;
    const schemaName = input.SchemaName;
    const schemaDefinition = input.SchemaDefinition;
    const dataFormat = input.DataFormat;

    const key = definitionKey(registryName, schemaName, schemaDefinition);
    // Idempotency without `Force*` guidance: real Glue rejects a duplicate
    // `CreateSchema` with `AlreadyExistsException`, so the fake does the
    // same when the schema name already exists in the store. A scenario
    // that wants to short-circuit this can set `forceCreateError` to a
    // named error directly.
    if (this.definitionToVersion.has(key)) {
      throw makeAlreadyExistsError(
        `schema already exists: schemaName=${schemaName ?? ""}`,
      );
    }

    const versionId = this.nextVersionId();
    const stored: StoredSchemaVersion = {
      schemaVersionId: versionId,
      schemaName: schemaName ?? "",
      registryName: registryName ?? "",
      schemaDefinition: schemaDefinition ?? "",
      dataFormat: (dataFormat as string | undefined) ?? "",
      compatibility: input.Compatibility as string | undefined,
      description: input.Description,
      tags: { ...(input.Tags ?? {}) },
      metadata: {},
      pendingCount: this.forceRegisterPending ? this.forcePendingCount : 0,
    };
    this.definitionToVersion.set(key, versionId);
    this.versionToSchema.set(versionId, stored);

    return {
      $metadata: {},
      SchemaVersionId: versionId,
      SchemaName: schemaName,
      SchemaArn: this.schemaArnFor(stored),
      DataFormat: dataFormat,
      Compatibility: input.Compatibility,
      Description: input.Description,
      Tags: input.Tags,
      // `CreateSchemaResponse` exposes the first version's status via
      // `SchemaVersionStatus` (not the generic `Status` used on
      // `RegisterSchemaVersion` / `GetSchemaByDefinition`).
      SchemaVersionStatus: this.statusFor(stored),
    };
  }

  async registerSchemaVersion(
    input: RegisterSchemaVersionCommandInput,
  ): Promise<RegisterSchemaVersionCommandOutput> {
    this.counts.registerSchemaVersion += 1;

    const registryName = input.SchemaId?.RegistryName;
    const schemaName = input.SchemaId?.SchemaName;
    const schemaDefinition = input.SchemaDefinition;

    const key = definitionKey(registryName, schemaName, schemaDefinition);
    // If the exact (name, definition) is already registered, real Glue
    // returns the existing SchemaVersionId with `AVAILABLE` (register is
    // idempotent on identical definitions). The fake mirrors that.
    const existing = this.definitionToVersion.get(key);
    if (existing !== undefined) {
      const stored = this.versionToSchema.get(existing);
      return {
        $metadata: {},
        SchemaVersionId: existing,
        Status: this.statusFor(stored),
      };
    }

    const versionId = this.nextVersionId();
    // Register-under-existing-schema does not carry DataFormat/Compatibility
    // on the input; carry over the fields from any prior version of the
    // same schema when available, else record an empty format.
    const priorFormat = this.findPriorDataFormat(registryName, schemaName);
    const stored: StoredSchemaVersion = {
      schemaVersionId: versionId,
      schemaName: schemaName ?? "",
      registryName: registryName ?? "",
      schemaDefinition: schemaDefinition ?? "",
      dataFormat: priorFormat ?? "",
      metadata: {},
      tags: {},
      pendingCount: this.forceRegisterPending ? this.forcePendingCount : 0,
    };
    this.definitionToVersion.set(key, versionId);
    this.versionToSchema.set(versionId, stored);

    return {
      $metadata: {},
      SchemaVersionId: versionId,
      Status: this.statusFor(stored),
    };
  }

  async putSchemaVersionMetadata(
    input: PutSchemaVersionMetadataCommandInput,
  ): Promise<PutSchemaVersionMetadataCommandOutput> {
    this.counts.putSchemaVersionMetadata += 1;

    const versionId = input.SchemaVersionId;
    const kv = input.MetadataKeyValue;
    if (versionId === undefined) {
      throw makeEntityNotFoundError("SchemaVersionId is required");
    }
    const stored = this.versionToSchema.get(versionId);
    if (stored === undefined) {
      throw makeEntityNotFoundError(
        `no schema version found: schemaVersionId=${versionId}`,
      );
    }
    if (kv?.MetadataKey === undefined) {
      return { $metadata: {}, SchemaVersionId: versionId };
    }
    // Duplicate-key writes surface `AlreadyExistsException` — the real Glue
    // behaviour the writer path treats as an idempotent success.
    if (Object.prototype.hasOwnProperty.call(stored.metadata, kv.MetadataKey)) {
      throw makeAlreadyExistsError(
        `metadata key already set: schemaVersionId=${versionId} key=${kv.MetadataKey}`,
      );
    }
    stored.metadata[kv.MetadataKey] = kv.MetadataValue ?? "";
    return {
      $metadata: {},
      SchemaVersionId: versionId,
      MetadataKey: kv.MetadataKey,
      MetadataValue: kv.MetadataValue,
    };
  }

  async querySchemaVersionMetadata(
    input: QuerySchemaVersionMetadataCommandInput,
  ): Promise<QuerySchemaVersionMetadataCommandOutput> {
    this.counts.querySchemaVersionMetadata += 1;

    const versionId = input.SchemaVersionId;
    if (versionId === undefined) {
      throw makeEntityNotFoundError("SchemaVersionId is required");
    }
    const stored = this.versionToSchema.get(versionId);
    if (stored === undefined) {
      throw makeEntityNotFoundError(
        `no schema version found: schemaVersionId=${versionId}`,
      );
    }
    // Flatten into the SDK's `MetadataInfoMap` shape — the reader consumes
    // `info.MetadataValue` off each entry.
    const infoMap: Record<string, { MetadataValue: string }> = {};
    for (const [k, v] of Object.entries(stored.metadata)) {
      infoMap[k] = { MetadataValue: v };
    }
    return {
      $metadata: {},
      SchemaVersionId: versionId,
      MetadataInfoMap: infoMap,
    };
  }

  async getTags(input: GetTagsCommandInput): Promise<GetTagsCommandOutput> {
    this.counts.getTags += 1;

    const arn = input.ResourceArn;
    if (arn === undefined) {
      return { $metadata: {}, Tags: {} };
    }
    // The fake tracks tags on the primary (first) version of a schema (Glue
    // stores tags on the schema resource, not per version); scan by ARN.
    for (const stored of this.versionToSchema.values()) {
      if (this.schemaArnFor(stored) === arn) {
        return { $metadata: {}, Tags: { ...stored.tags } };
      }
    }
    return { $metadata: {}, Tags: {} };
  }

  // --- Internal helpers --------------------------------------------------

  /**
   * Report `AVAILABLE` when the version has no remaining pending polls,
   * else `PENDING`. Matches Glue's status vocabulary; the seven other
   * statuses (`DELETING` / `FAILURE` / etc.) are not modelled here because
   * a scenario that needs one drives it explicitly via `Force*`.
   */
  private statusFor(
    stored: StoredSchemaVersion | undefined,
  ): "AVAILABLE" | "PENDING" | undefined {
    if (stored === undefined) return undefined;
    return stored.pendingCount > 0 ? "PENDING" : "AVAILABLE";
  }

  /**
   * Synthesise a schema ARN in the Glue-idiomatic shape so tag lookup by
   * ARN can round-trip. Not a real ARN — the fake ignores account/region
   * and derives them from the registry name.
   */
  private schemaArnFor(stored: StoredSchemaVersion | undefined): string {
    if (stored === undefined) return "";
    return `arn:aws:glue:us-east-2:000000000000:schema/${stored.registryName}/${stored.schemaName}`;
  }

  /**
   * Return the `DataFormat` recorded on the earliest version of a given
   * `(registryName, schemaName)` pair, or `undefined` if none exists.
   * `RegisterSchemaVersion` does not carry a `DataFormat` on its input, so
   * a follow-on register needs the format from a prior version to answer a
   * later `getSchemaVersion` call faithfully.
   */
  private findPriorDataFormat(
    registryName: string | undefined,
    schemaName: string | undefined,
  ): string | undefined {
    const rn = registryName ?? "";
    const sn = schemaName ?? "";
    for (const stored of this.versionToSchema.values()) {
      if (stored.registryName === rn && stored.schemaName === sn) {
        return stored.dataFormat;
      }
    }
    return undefined;
  }

  /**
   * Hand out the next `SchemaVersionId`. Seeded UUIDs from the constructor
   * are consumed first; when exhausted, generate a deterministic UUID from
   * a monotonic counter so tests stay reproducible without depending on
   * `crypto.randomUUID`.
   */
  private nextVersionId(): string {
    const seeded = this.seededVersionIds.shift();
    if (seeded !== undefined) {
      return seeded;
    }
    const n = this.nextGeneratedId++;
    const hex = n.toString(16).padStart(12, "0");
    return `00000000-0000-0000-0000-${hex}`;
  }
}
